dtfs-central-api/src/services/portal/deal.service.ts

```diff
@@ -40,4 +41,32 @@ export class PortalDealService {
       throw new Error(`Invalid dealType ${dealType}`);
     }
   }
+
+  /**
+   * Add a "deal cancelled" activity
+   *
+   * @param addDealCancelledActivity
+   * @param addDealCancelledActivity.dealId - the deal Id to update
+   * @param addDealCancelledActivity.dealType - the deal type
+   * @param addDealCancelledActivity.portalActivities - previous/existing deal activities
+   * @param addDealCancelledActivity.author - the activity's author
+   * @param addDealCancelledActivity.auditDetails - the users audit details
+   */
+  public static async addDealCancelledActivity({
+    dealId,
+    dealType,
+    portalActivities,
+    author,
+    auditDetails,
+  }: {
+    dealId: ObjectId | string;
+    dealType: DealType;
+    portalActivities: Array<Activity>;
+    author: ActivityAuthor;
+    auditDetails: AuditDetails;
+  }): Promise<void> {
+    if (dealType === DEAL_TYPE.GEF) {
+      await addGefDealCancelledActivity({ dealId, portalActivities, author, auditDetails });
```

Comment:
In other activitiy logic, `portalActivities` is also passed around so stuck with the same pattern.

We can't use `$push` because - without reactoring existing activities logic, the latest activity needs to be the first item in the activities array. Otherwise the design will not match and activity patterns would be misaligned

Reply:
It looks like you can use `$position` alongside `$push` https://www.mongodb.com/docs/manual/reference/operator/update/push/

It should be fine but theres definitely a risk of a race condition/overwriting the existing activities.


I think my preference would be to create a new `addPortalActivity` repository method & use 
```
collection.update({ dealId }, {
  $push: {
    activities: {
      $each: [{ newActivity }],
      $position: 0
    }
  }
})
```

Reply:
Great shout @MarRobSoftwire - I was following the existing activity approaches which carry the risk you mention.

`$position` Mongo method is new to me - nice 👍 

I am creating a portal activity repo to do this. Will also add a ticket to the backlog to update other activity logic to use the repo.